include/oneapi/dpl/internal/dynamic_selection.h

```diff
@@ -0,0 +1,228 @@
+        // policy has optional submit_and_wait(selection, f, args...)
+        return t.get_policy().submit_and_wait(std::forward<T>(t), std::forward<Function>(f), std::forward<Args>(args)...);
+      } else {
+        // policy does not have optional submit_and_wait for a selection
+        return wait(submit(std::forward<T>(t), std::forward<Function>(f), std::forward<Args>(args)...));
+      }
+    } else {
+      // t is a policy
+      if constexpr ( internal::has_submit_and_wait<T, Function, Args...>::value == true) {
+        // has the optional submit_and_wait(f, args...)
+        return std::forward<T>(t).submit_and_wait(std::forward<Function>(f), std::forward<Args>(args)...);
+      } else if constexpr ( internal::has_submit_and_wait_handle<T, typename std::decay<T>::type::selection_type, Function, Args...>::value == true) {
+        // has the optional submit_and_wait for a selection, so select and call
+        return submit_and_wait(std::forward<T>(t).select(f, args...), std::forward<Function>(f), std::forward<Args>(args)...);
+      } else {
+        // does not have the optional submit_and_wait(f, args...) or (s, f, args...)
+        return wait(submit(std::forward<T>(t), std::forward<Function>(f), std::forward<Args>(args)...));
+      }
+    }
+  }
+
+  // support for execution info
+
+  namespace execution_info {
+    struct task_time_t {
+      static constexpr bool is_execution_info_v = true;
+      using value_type = uint64_t;
+    };
+    inline constexpr task_time_t task_time;
+
+    struct task_completion_t {
+      static constexpr bool is_execution_info_v = true;
+      using value_type = void;
+    };
+    inline constexpr task_completion_t task_completion;
+
+    struct task_submission_t {
+      static constexpr bool is_execution_info_v = true;
+      using value_type = void;
+    };
+    inline constexpr task_submission_t task_submission;
+  } // namespace execution_info
+
+  template<typename S, typename Info>
+  void report(S&& s, const Info& i) {
+    if constexpr(internal::has_report<S,Info>::value == true) {
+      std::forward<S>(s).report(i);
+    }
+  }
+
+
+  template<typename S, typename Info, typename Value>
+  void report(S&& s, const Info& i, const Value& v) {
+    if constexpr(internal::has_report_value<S,Info>::value == true) {
+      std::forward<S>(s).report(i, v);
+    }
+  }
+
+  template<typename S, typename Info>
+  struct report_info {
+    static constexpr bool value = internal::has_report<S,Info>::value;
+  };
+  template<typename S, typename Info>
+  inline constexpr bool report_info_v = report_info<S,Info>::value;
+
+  template<typename S, typename Info>
+  struct report_value {
+    static constexpr bool value = internal::has_report_value<S,Info>::value;
+  };
+  template<typename S, typename Info>
+  inline constexpr bool report_value_v = report_value<S,Info>::value;
+
+} // namespace experimental
+} // namespace dpl
+} // namespace oneapi
+
+#if _DS_BACKEND_SYCL != 0
```

Comment:
Am I understanding your comment correctly, that you think sycl_backend.h should be included by each policy, instead of here?  And that which file should be dynamic_selection_traits -- all of dynamic_selection.h?  Can you be a bit more specific about what you are proposing?

Reply:
Header issues fixed with introduction of dynamic_selection_traits.h and review of missing, unnecessary includes.